### ISSUE TITLE
Reinstantiate automatic use of r.json

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#3843]
We had another user running into a case where sending large documents to the server was extremely slow. It becomes much faster when wrapping the document into `r.json`, like we used to do before we switched away from protocol buffers.

We should put that logic back in, at least until we can fix https://github.com/rethinkdb/rethinkdb/issues/2900 which might be the reason for why things are slow without `r.json`.

This might also solve https://github.com/rethinkdb/rethinkdb/issues/2919 (depending on whether we use `r.json` automatically for `setUnion` arguments or just for `insert`/`replace` etc.).

Would be nice to get it into 2.0
(@Tryneus, @mlucy, @VeXocide any takers?)

**Comment by neumino** [comment]
I'm not sure that `r.json` is a good solution. The (removed) method `exprJSON` in the JS driver comes at a certain cost. You basically have to make sure that there's no ReQL term in your data and then call `r.json(JSON.stringify(...))`.

It used to be cheaper to do that than to build the protobuf object, but now that we don't have to create a datum for each value (except arrays), it's not obvious that this is more efficient. It's especially not in Node when the object has too many fields and v8 switch to a hash table to represent it.

Having something like `r.value(<json value>)` where `<json value>` does not contain a ReQL term (like r.now(), r.random() etc.) and (`r.value(...) = r.json(JSON.stringify(...))` would make more sense.

You can then have `insert` by default wrap its first argument with `r.value` and add a flag if people want to insert documents with ReQL terms.

**Comment by neumino** [comment]
I actually run some tests this week end with rethinkdbdash. For a document `doc` with about 30k field, all strings, with a total size of about 1MB:

```js
r.table('data').insert(r.json(docStr)) // ~10ms for the driver
r.table('data').insert(r.json(JSON.stringify(doc))) // ~300ms for the driver
r.table('data').insert(doc) // ~500ms for the driver
```

And this also takes 500 ms (the loop just checks if all keys are an instance of something, which is similar to what `r.exprJSON` would have to do):
```
function Foo() {}
var keys = Object.keys(doc);
for(var i=0, len=keys.length; i<len; i++) {
    if (doc[keys[i]] instanceof Foo) { process.exit(1) }
}
r.table('data').insert(r.json(JSON.stringify(doc)))
```

It may work better in Python/Ruby as this may be tied to the fact that v8 doesn't handle well big objects. But it's not obviously better to re-instantiate `r.exprJSON` at least in JavaScript.

**Comment by danielmewes** [comment]
Interesting benchmarks @neumino .
Using `r.json` definitely makes the server use less resources, but as you've shown that might be offset by higher costs on the client.

It sounds like we need to give this more consideration. In the meantime we can tell users who deal with large nested documents to use `r.json(JSON.stringify(doc))` or something equivalent to that in their respective language.

Once we fix https://github.com/rethinkdb/rethinkdb/issues/2900 the problem will hopefully go away.

**Comment by danielmewes** [comment]
Moving out of 2.0

**Comment by danielmewes** [comment]
Let's just fix #2900.
